recipes/full_finetune_distributed.py

```diff
@@ -637,11 +776,17 @@ def save_checkpoint(
 
         intermediate_checkpoint = epoch + 1 < self.total_epochs
 
+        # If async checkpointing is enabled, intermediate checkpoints will
+        # be saved asynchronously.
+        if intermediate_checkpoint and self._enable_async_checkpointing:
+            self.save_checkpoint_async(epoch)
```

Comment:
i like that we have a function specialized for async. This would get a bit into refactoring, but it would be great to take the other part of the code and turn it into self.save_checkpoint_sync.

Reply:
New commit refactors the checkpointing code into its own client and cleans up the checkpointing code from the recipes. I believe this comment should already be addressed. Post refactor, we have distinct methods for sync and async saves. PTAL.